working/closed-and-base/feature-specification.md

```diff
@@ -0,0 +1,84 @@
+# "Closed" and "base" types
+
+Author: Bob Nystrom
+
+Status: In-progress
+
+Version 1.0
+
+This proposal specifies `closed` and `base` modifiers on classes and mixins,
+which allow an author to prohibit the type being extended or implemented,
+respectively.
+
+This proposal is a subset of the [type modifiers][] strawman, which also
+contains most of the motivation. It is split out here because the type modifiers
+strawman also proposes prohibiting classes being used as mixins, which is a
+larger breaking change. This proposal is non-breaking.
+
+[type modifiers]: https://github.com/dart-lang/language/blob/master/working/type-modifiers/feature-specification.md
+
+## Syntax
+
+A class declaration may be preceded with the built-in identifiers
+`closed` and/or `base`:
+
+```
+classDeclaration ::=
+  'closed'? 'abstract'? 'base'? 'class' identifier typeParameters?
+  superclass? interfaces?
+  '{' (metadata classMemberDeclaration)* '}'
+  | 'closed'? 'abstract'? 'base'? 'class' mixinApplicationClass
+```
+
+A mixin declaration may be preceded with the built-in identifier `base`:
+
+```
+mixinDeclaration ::= 'base'? 'mixin' identifier typeParameters?
+  ('on' typeNotVoidList)? interfaces?
+  '{' (metadata classMemberDeclaration)* '}'
+```
+
+This proposal will likely build on top of the [sealed types][] proposal, in
+which case the full grammar is:
+
+```
+classDeclaration ::=
+  'closed'? ('sealed' | 'abstract')? 'base'? 'class' identifier typeParameters?
```

Comment:
Ah, good point! If we *don't* have `closed` and `base`, then it's useful to allow `sealed` even on types with no subtypes because it gives you a backdoor way to express `closed base`. But if *do* have those, as this proposes, then it's redundant to put `closed` and `base` on a class also marked `sealed`.

It doesn't conflict because even when you mark a class `closed` and/or `base` you can still extend/implement it *in the same library. But it is redundant, and we should probably prohibit it. Updated.

Reply:
I don't think it's redundant, because of the transitiveness of base, right?